docs/spec.md

```diff
@@ -2437,8 +2440,8 @@ the XLA compiler. In the future, we are planning to unify this metadata
 %results = "stablehlo.custom_call"(%input0) {
   call_target_name = "foo",
   has_side_effect = false,
-  backend_config = "bar",
-  api_version = 1 : i32,
+  backend_config = {bar = 42 : i32},
```

Comment:
I'm worried that this is going to break a lot of stuff. What will our approach be for compatibility? Should we introduce a new attribute and deprecate backend_config?

Reply:
Could you elaborate? I don't think this poses any compatibility risks, like with any new feature, if you use the new feature it isn't forward compatible. And for backward_compat, <APIv4,dictionary> is an option, and <APIv3, string> will still work as well.

Reply:
Sorry, I didn't realize that the `api_version` and `backend_config` fields were related. Maybe this should be explained in the spec and in the RFC? I agree in this case there is no concern as we are merely adding a new feature.

(My concern was that I thought we were going to just switch completely from strings to dictionaries. That would have been a breaking change, and it would have been unclear how to deserialize old bytecode.)

Reply:
spec says : "`backend_config` and `api_version` may be used to provide additional implementation-defined metadata"  so don't need change here? but  I will add more details in RFC to make it clear.

Reply:
IMO it would be helpful to provide more information on what the `api_version` and `backend_config` are intended for and how they relate to each other. I think "backend_config and api_version may be used to provide additional implementation-defined metadata" is too broad. In MHLO there is quite a bit more information: https://github.com/tensorflow/tensorflow/blob/54a599afec6e0fc4afb48385e0cd4e0b8e512cd0/third_party/xla/xla/mlir_hlo/mhlo/IR/hlo_ops.td#L2514-L2526

Reply:
that information will be added to the op def, not sure about the spec considering those are implementation details. Also, if we talk about `API_VERSION_TYPED_FFI` in the spec we will need to provide details for all `api_version`s ?

@GleasonK  / @sdasgup3  what is your opinion?

Reply:
IMO this information is relevant to have in both places. I do think this belongs in the spec given that the value of `api_version` influences the meaning of `backend_config`.

Curious to hear what others think!

Reply:
I totally agree that the section in the spec which talks about `backend_config`, `api_version`  are  vague and not an ideal state to be in. But I would say that it is "intentional". That is mainly because at the point of writing the section there were various unknowns like https://github.com/tensorflow/mlir-hlo/blob/e7b1891ba261c2ab1e74b6979d16fd6d648a1232/mhlo/IR/hlo_ops.td#L2534 and #741.  At some point we thought about enumerating all the enums for `api_version`, but backed out for reasons like https://github.com/tensorflow/mlir-hlo/blob/e7b1891ba261c2ab1e74b6979d16fd6d648a1232/mhlo/IR/hlo_ops_enums.td#L110 (mainly) and lack on proper documentation on available enums which would have forced us to write something vague as we did in https://github.com/openxla/stablehlo/blob/main/docs/spec.md#rng_bit_generator. The outcome we decided was call out "backend_config" and "apli_version" as measn to provide implementation defined metadata together with #741.

My proposal: How about we keep the current state and in very near future  focus on  #741 to resolve the open questions.